src/server/search/search_family.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright 2022, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/search/search_family.h"
+
+#include <atomic>
+#include <jsoncons/json.hpp>
+#include <variant>
+#include <vector>
+
+#include "base/logging.h"
+#include "core/json_object.h"
+#include "core/search/search.h"
+#include "facade/error.h"
+#include "facade/reply_builder.h"
+#include "server/command_registry.h"
+#include "server/conn_context.h"
+#include "server/container_utils.h"
+#include "server/engine_shard_set.h"
+#include "server/search/search_index.h"
+#include "server/transaction.h"
+
+namespace dfly {
+
+using namespace std;
+using namespace facade;
+
+void SearchFamily::FtCreate(CmdArgList args, ConnectionContext* cntx) {
+  string_view idx_name = ArgS(args, 0);
+
+  SearchIndex index{};
+
+  for (size_t i = 1; i < args.size(); i++) {
+    ToUpper(&args[i]);
+
+    // [ON HASH | JSON]
+    if (ArgS(args, i) == "ON") {
+      if (++i >= args.size())
+        return (*cntx)->SendError(kSyntaxErr);
+
+      ToUpper(&args[i]);
+      string_view type = ArgS(args, i);
+      if (type == "HASH")
+        index.type = SearchIndex::HASH;
+      else if (type == "JSON")
+        index.type = SearchIndex::JSON;
+      else
+        return (*cntx)->SendError("Invalid rule type: " + string{type});
+      continue;
+    }
+
+    // [PREFIX count prefix [prefix ...]]
+    if (ArgS(args, i) == "PREFIX") {
+      if (i + 2 >= args.size())
+        return (*cntx)->SendError(kSyntaxErr);
+
+      if (ArgS(args, ++i) != "1")
+        return (*cntx)->SendError("Multiple prefixes are not supported");
+
+      index.prefix = ArgS(args, ++i);
+      continue;
+    }
+  }
+
+  auto idx_ptr = make_shared<SearchIndex>(move(index));
+  cntx->transaction->ScheduleSingleHop([idx_name, idx_ptr](auto* tx, auto* es) {
+    ShardSearchIndex::InitOnShard(tx->GetOpArgs(es), idx_name, idx_ptr);
+    return OpStatus::OK;
+  });
+
+  (*cntx)->SendOk();
+}
+
+void SearchFamily::FtSearch(CmdArgList args, ConnectionContext* cntx) {
+  string_view index_name = ArgS(args, 0);
+  string_view query_str = ArgS(args, 1);
+
+  optional<search::SearchAlgorithm> search_algo{};
+  try {
```

Comment:
I avoid using c++ exceptions in general unless I do not have any options. 
In this case, seems you can just introduce an `Init` function.